File: apps/api/src/investhome_api/schemas/crm_matches.py
"""CRM identity match review schemas — no fake scores, no merge side effects."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

CrmMatchKind = Literal["strong", "possible"]
CrmMatchStatus = Literal["pending", "in_review", "same_person", "different"]
CrmMatchAction = Literal["same_person", "different", "in_review"]
CrmMatchReason = Literal["phone", "email", "secondary_email", "bitrix", "name_review"]


class CrmMatchEvidence(BaseModel):
    reason: str
    value: str | None = None


class CrmMatchPurchase(BaseModel):
    id: UUID
    project: str | None = None
    unit: str | None = None
    source_id: str | None = None
    historical: bool = False


class CrmMatchPerson(BaseModel):
    id: UUID
    name: str
    phone: str | None = None
    email: str | None = None
    secondary_emails: list[str] = Field(default_factory=list)
    source: str | None = None
    source_ids: list[str] = Field(default_factory=list)
    notes: str | None = None
    review_required: bool = False
    href: str
    purchases: list[CrmMatchPurchase] = Field(default_factory=list)


class CrmMatchItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    person_a_id: UUID
    person_a_name: str
    person_b_id: UUID
    person_b_name: str
    reasons: list[str] = Field(default_factory=list)
    match_kind: CrmMatchKind
    shared_phone: str | None = None
    shared_email: str | None = None
    source: str | None = None
    status: CrmMatchStatus
    protected: bool = False
    protected_reason: str | None = None
    merge_queued: bool = False
    review_notes: str | None = None
    created_at: datetime | None = None


class CrmMatchDetail(CrmMatchItem):
    person_a: CrmMatchPerson
    person_b: CrmMatchPerson
    evidence: list[CrmMatchEvidence] = Field(default_factory=list)
    conflicts: list[str] = Field(default_factory=list)
    linked_purchases: list[CrmMatchPurchase] = Field(default_factory=list)
    merge_blocked: bool = True
    merge_message: str


class CrmMatchKpis(BaseModel):
    pending: int = 0
    strong: int = 0
    possible: int = 0
    rejected: int = 0
    same_person: int = 0


class CrmMatchListResponse(BaseModel):
    items: list[CrmMatchItem]
    total: int
    kpis: CrmMatchKpis
    sources: list[str] = Field(default_factory=list)
    reasons: list[str] = Field(default_factory=list)


class CrmMatchDecisionRequest(BaseModel):
    action: CrmMatchAction
    notes: str | None = Field(default=None, max_length=4000)

File: apps/web/src/workspaces/crm/api/crm-matches.ts
import { apiFetch } from '@/lib/api/client';

export type CrmMatchKind = 'strong' | 'possible';
export type CrmMatchStatus = 'pending' | 'in_review' | 'same_person' | 'different';
export type CrmMatchAction = 'same_person' | 'different' | 'in_review';

export type CrmMatchEvidence = {
  reason: string;
  value?: string | null;
};

export type CrmMatchPurchase = {
  id: string;
  project?: string | null;
  unit?: string | null;
  source_id?: string | null;
  historical: boolean;
};

export type CrmMatchPerson = {
  id: string;
  name: string;
  phone: string | null;
  email: string | null;
  secondary_emails: string[];
  source: string | null;
  source_ids: string[];
  notes: string | null;
  review_required: boolean;
  href: string;
  purchases: CrmMatchPurchase[];
};

export type CrmMatchItem = {
  id: string;
  person_a_id: string;
  person_a_name: string;
  person_b_id: string;
  person_b_name: string;
  reasons: string[];
  match_kind: CrmMatchKind;
  shared_phone: string | null;
  shared_email: string | null;
  source: string | null;
  status: CrmMatchStatus;
  protected: boolean;
  protected_reason: string | null;
  merge_queued: boolean;
  review_notes: string | null;
  created_at?: string | null;
};

export type CrmMatchDetail = CrmMatchItem & {
  person_a: CrmMatchPerson;
  person_b: CrmMatchPerson;
  evidence: CrmMatchEvidence[];
  conflicts: string[];
  linked_purchases: CrmMatchPurchase[];
  merge_blocked: boolean;
  merge_message: string;
};

export type CrmMatchKpis = {
  pending: number;
  strong: number;
  possible: number;
  rejected: number;
  same_person: number;
};

export type CrmMatchListResponse = {
  items: CrmMatchItem[];
  total: number;
  kpis: CrmMatchKpis;
  sources: string[];
  reasons: string[];
};

export type CrmMatchListParams = {
  search?: string;
  match_type?: string;
  status?: string;
  source?: string;
};

function buildParams(values: Record<string, string | undefined>): URLSearchParams {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(values)) {
    if (!value) continue;
    params.set(key, value);
  }
  return params;
}

export async function fetchCrmMatches(params: CrmMatchListParams = {}): Promise<CrmMatchListResponse> {
  const query = buildParams({
    search: params.search,
    match_type: params.match_type,
    status: params.status,
    source: params.source,
  });
  const suffix = query.toString() ? `?${query.toString()}` : '';
  return apiFetch<CrmMatchListResponse>(`/crm/matches${suffix}`);
}

export async function fetchCrmMatch(id: string): Promise<CrmMatchDetail> {
  return apiFetch<CrmMatchDetail>(`/crm/matches/${id}`);
}

export async function decideCrmMatch(
  id: string,
  action: CrmMatchAction,
  notes?: string,
): Promise<CrmMatchDetail> {
  return apiFetch<CrmMatchDetail>(`/crm/matches/${id}/decision`, {
    method: 'POST',
    body: JSON.stringify({ action, notes: notes || null }),
  });
}
